### PR TITLE
Document row pattern recognition in window

### PR DESCRIPTION
**Pull request by kasiafi**

### CODE REVIEW

**Comment by kasiafi** [comment]
It would be good to merge this before the upcoming release, because the related change is already in.

**Comment by kasiafi** [comment]
@martint @mosabua please review.

**Review by mosabua — CHANGES_REQUESTED** [review]
First of all - nice work again as usual!

I just did a first pass for now. Here are a couple of general remarks beyond the specific comments:

* hard to figure out when we use "window" and when "window structure" 
* saying that it is same as in match_recognize is maybe not ideal .. should we replicate the content and adjust?
* do we need some more examples since this is all pretty complicated, like some simple and some more complex ones?

Ping me if you want to chat.

File: docs/src/main/sphinx/sql/pattern-recognition-in-window.rst [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+window the pattern matching can neither match rows nor retrieve input values
+outside the frame.
+
+Besides the ``frame_extent``, the two inevitable clauses for pattern matching
+are ``PATTERN`` and ``DEFINE``.
+
+The ``PATTERN`` clause specifies a row pattern which is a form of a regular
+expression with some syntactical extensions. The row pattern syntax is similar
+to the :ref:`row pattern syntax in MATCH_RECOGNIZE<row_pattern_syntax>`.
+However, the anchor pattern: ``^``, ``$`` is not allowed in window.
```
[the remark this answers is not in the document]
Comment (kasiafi) [reply docs/src/main/sphinx/sql/pattern-recognition-in-window.rst:84]: Yes, there are two. I separated them with "or" to make it clear.

File: docs/src/main/sphinx/sql/pattern-recognition-in-window.rst [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+                     PATTERN (A B+ C+)
+                     DEFINE
+                            B AS B.value < PREV (B.value),
+                            C AS C.value > PREV (C.value)
+                    )
+
+Measures defined in a window can be called in the ``SELECT`` clause and in the
+``ORDER BY`` clause of the enclosing query. The ``RUNNING`` and ``FINAL``
+keywords are allowed, however they have no effect, since the semantics of
+the computations is effectively final.
```
[the remark this answers is not in the document]
Comment (kasiafi) [reply docs/src/main/sphinx/sql/pattern-recognition-in-window.rst:119]: I added a separate paragraph about `RUNNING` and `FINAL`.

**Comment by kasiafi** [comment]
@mosabua thanks for the review! I applied your comments in a separate commit.

Regarding the top-level comments:
> hard to figure out when we use "window" and when "window structure"

I'd refer to the syntax as "window structure", and to the runtime abstraction as "window"

> saying that it is same as in match_recognize is maybe not ideal .. should we replicate the content and adjust?

I assume that `MATCH_RECOGNIZE` is the basic / most popular way to do row pattern matching. I think that pattern matching in window is used by the people who are already familiar with the concept, so they are most likely to read the docs for the differences rather than for the full explanation.

> do we need some more examples since this is all pretty complicated, like some simple and some more complex ones?

I could add examples of particular clauses.

File: docs/src/main/sphinx/sql/pattern-recognition-in-window.rst [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+
+For every input row processed by the window, the portion of rows enclosed by
+the ``frame_extent`` limits the search area for row pattern recognition. Unlike
+in ``MATCH_RECOGNIZE``, where the pattern search can explore all rows until the
+partition end, and all rows of the partition are available for computations, in
+window structures the pattern matching can neither match rows nor retrieve
+input values outside the frame.
+
+Besides the ``frame_extent``, the two required clauses for pattern matching are
+``PATTERN`` and ``DEFINE``.
```
Comment (martint) [inline docs/src/main/sphinx/sql/pattern-recognition-in-window.rst:82]: Besides the ``frame_extent``, pattern matching requires the ``PATTERN`` and ``DEFINE`` clauses.

File: docs/src/main/sphinx/sql/pattern-recognition-in-window.rst [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+
+The ``PATTERN`` clause specifies a row pattern which is a form of a regular
```
Comment (martint) [inline docs/src/main/sphinx/sql/pattern-recognition-in-window.rst:84]: "... row pattern, which is ..."

File: docs/src/main/sphinx/sql/pattern-recognition-in-window.rst [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+expression with some syntactical extensions. The row pattern syntax is similar
+to the :ref:`row pattern syntax in MATCH_RECOGNIZE<row_pattern_syntax>`.
+However, the anchor pattern: ``^`` or ``$`` is not allowed in window.
```
Comment (martint) [inline docs/src/main/sphinx/sql/pattern-recognition-in-window.rst:87]: "However, the anchor patterns ``^`` and ``$`` are not allowed in a window specification."

File: docs/src/main/sphinx/sql/pattern-recognition-in-window.rst [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+
+The ``DEFINE`` clause is where row pattern primary variables are defined by
+boolean conditions. It is similar to the
```
Comment (martint) [inline docs/src/main/sphinx/sql/pattern-recognition-in-window.rst:90]: The ``DEFINE`` clause defines the row pattern primary variables in terms of boolean conditions that must be satisfied.

File: docs/src/main/sphinx/sql/pattern-recognition-in-window.rst [hunk 7]
```diff
@@ -0,0 +1,249 @@
...
+:ref:`DEFINE clause of MATCH_RECOGNIZE<row_pattern_variable_definitions>`.
+The only difference is that the window syntax does not support the
+``MATCH_NUMBER`` function.
+
+The ``MEASURES`` clause is syntactically similar to the
+:ref:`MEASURES clause of MATCH_RECOGNIZE<row_pattern_measures>`, the only
```
Comment (martint) [inline docs/src/main/sphinx/sql/pattern-recognition-in-window.rst:96]: Period before "the only"

File: docs/src/main/sphinx/sql/pattern-recognition-in-window.rst [hunk 8]
```diff
@@ -0,0 +1,249 @@
...
+                            RUNNING LAST(total_price) AS value,
+                            CLASSIFIER() AS label
+                     ROWS BETWEEN CURRENT ROW AND UNBOUNDED FOLLOWING
+                     PATTERN (A B+ C+)
+                     DEFINE
+                            B AS B.value < PREV (B.value),
+                            C AS C.value > PREV (C.value)
+                    )
+
+Measures defined in a window can be called in the ``SELECT`` clause and in the
```
Comment (martint) [inline docs/src/main/sphinx/sql/pattern-recognition-in-window.rst:119]: "can be called" -> "can be referenced"

File: docs/src/main/sphinx/sql/pattern-recognition-in-window.rst [hunk 9]
```diff
@@ -0,0 +1,249 @@
...
+``ORDER BY`` clause of the enclosing query.
+
+The ``RUNNING`` and ``FINAL`` keywords are allowed in the ``MEASURES`` clause.
+They can precede a logical navigation function ``FIRST`` or ``LAST``. However,
+they have no effect. Every computation is performed from the position of the
+final row of the match, so the semantics is effectively ``final``.
```
Comment (martint) [inline docs/src/main/sphinx/sql/pattern-recognition-in-window.rst:125]: Capitalize FINAL.

File: docs/src/main/sphinx/sql/pattern-recognition-in-window.rst [hunk 10]
```diff
@@ -0,0 +1,249 @@
...
+              output
+            * Evaluate the ``AFTER MATCH SKIP`` clause, and mark the 'skipped'
+              rows
+
+Empty matches and unmatched rows
+--------------------------------
+
+If no match can be associated with a particular input row, the row is
+*unmatched*. This happens when no match can be found for the row. This also
+happens when no match is attempted for the row, because it is ``skipped`` by
```
Comment (martint) [inline docs/src/main/sphinx/sql/pattern-recognition-in-window.rst:221]: Why is skipped in ``?

File: docs/src/main/sphinx/sql/select.rst [hunk 11]
```diff
@@ -668,11 +668,18 @@ A window specification has the following components:
...
+  UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING``. The window frame syntax
+  supports additional clauses for row pattern recognition. If the row pattern
+  recognition clauses are specified, the window frame for a particular row
+  consists of the rows matched by a pattern starting from that row.
+  Additionally, if the frame specifies row pattern measures, they can be
+  called over the window, similarly to window functions. For more details, see
+  :doc:`Row pattern recognition in window structures
+  </sql/pattern-recognition-in-window>`.
+
+Each window component is optional. Defaults are used when window partitioning,
```
Comment (martint) [inline docs/src/main/sphinx/sql/select.rst:680]: The sentence starting at "Defaults ..." is hard to parse. Does the "and" associate more strongly than the last "or". Maybe split up the sentence:

"Defaults are used when the existing window name is absent and the window partitioning, ordering or frame are not specified. They are also used when ..."

Or make them a bullet list:

```
Defaults are used in any of these scenarios:
* ....
* ....
```

By the way, what does "existing window name is absent" mean?

Comment (kasiafi) [reply docs/src/main/sphinx/sql/select.rst]: "existing window name is absent" refers to the `existing window name` being part of window specification.

I tried to reword this paragraph, but might need more help with it.

File: docs/src/main/sphinx/sql/select.rst [hunk 12]
```diff
@@ -677,9 +677,12 @@ A window specification has the following components:
...
 
-Each window component is optional. Defaults are used when window partitioning,
-ordering or frame is not specified, and existing window name is absent, or the
-referenced window specification does not provide particular components.
+Each window component is optional. If a window specification does not specify
+window partitioning, ordering or frame, those components are obtained from
+the window specification referenced by the ``existing window name``, or from
+another window specification in the reference chain. In case when there is no
+``existing window name`` specified, or none of the referenced window
+specifications contains the component, default is used.
```
Comment (martint) [inline docs/src/main/sphinx/sql/select.rst:685]: "the default value is used"

File: docs/src/main/sphinx/sql/pattern-recognition-in-window.rst [hunk 13]
```diff
@@ -78,23 +78,24 @@ partition end, and all rows of the partition are available for computations, in
...
+boolean conditions that must be satisfied. It is similar to the
 :ref:`DEFINE clause of MATCH_RECOGNIZE<row_pattern_variable_definitions>`.
 The only difference is that the window syntax does not support the
 ``MATCH_NUMBER`` function.
 
 The ``MEASURES`` clause is syntactically similar to the
-:ref:`MEASURES clause of MATCH_RECOGNIZE<row_pattern_measures>`, the only
-limitation being the usage of the ``MATCH_NUMBER`` function. However, the
+:ref:`MEASURES clause of MATCH_RECOGNIZE<row_pattern_measures>`. The only
+limitation is the usage of the ``MATCH_NUMBER`` function. However, the
```
Comment (martint) [inline docs/src/main/sphinx/sql/pattern-recognition-in-window.rst:98]: Now that I re-read this, it's not clear what the limitation is. Maybe make it more explicit by saying "The only limitation is that the ``MATCH_NUMBER`` function is not allowed", or something of that sort.

**Review by martint — APPROVED** [review]
